pkg/virt-handler/selinux/context_executor.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package selinux
+
+import (
+	"fmt"
+	"os/exec"
+	"runtime"
+	"syscall"
+
+	"github.com/opencontainers/selinux/go-selinux"
+)
+
+type SELinuxContextExecutor struct {
+	cmdToExecute  *exec.Cmd
+	desiredLabel  string
+	originalLabel string
+	pid           int
+}
+
+func NewSELinuxContextExecutor(pid int, cmd *exec.Cmd) (*SELinuxContextExecutor, error) {
+	desiredLabel, err := getSelinuxLabelForPID(pid)
+	if err != nil {
+		return nil, err
+	}
+	originalLabel, err := getOriginalSelinuxLabel()
+	if err != nil {
+		return nil, err
+	}
+	return &SELinuxContextExecutor{
+		pid:           pid,
+		cmdToExecute:  cmd,
+		desiredLabel:  desiredLabel,
+		originalLabel: originalLabel,
+	}, nil
+}
+
+func (ce SELinuxContextExecutor) Execute() error {
+	if isSELinuxEnabled() {
+		if err := ce.setDesiredContext(); err != nil {
+			return err
+		}
+	}
+	defer ce.resetContext()
+
+	preventFDLeakOntoChild()
+	if err := ce.cmdToExecute.Run(); err != nil {
+		return fmt.Errorf("failed to execute command in launcher namespace %d: %v", ce.pid, err)
+	}
+	return nil
+}
+
+func (ce SELinuxContextExecutor) setDesiredContext() error {
+	runtime.LockOSThread()
+	if err := selinux.SetExecLabel(ce.desiredLabel); err != nil {
+		return fmt.Errorf("failed to switch selinux context to %s. Reason: %v", ce.desiredLabel, err)
+	}
+	return nil
+}
+
+func (ce SELinuxContextExecutor) resetContext() error {
+	err := selinux.SetExecLabel(ce.originalLabel)
+	runtime.UnlockOSThread()
+	return err
+}
+
+func isSELinuxEnabled() bool {
+	_, selinuxEnabled, err := NewSELinux()
+	return err == nil && selinuxEnabled
+}
+
+func getSelinuxLabelForPID(pid int) (string, error) {
+	launcherSELinuxLabel, err := selinux.FileLabel(fmt.Sprintf("/proc/%d/attr/current", pid))
+	if err != nil {
+		return "", fmt.Errorf("could not retrieve pid %d selinux label: %v", pid, err)
+	}
+	return launcherSELinuxLabel, nil
+}
+
+func getOriginalSelinuxLabel() (string, error) {
+	launcherSELinuxLabel, err := selinux.FileLabel("/proc/thread-self/attr/current")
```

Comment:
This fails on kind - works on the others.

@EdDev 2 options:
  - use the hardcoded `spc_t` selinux context
  - try to do this, if the file doesn't exist, fall back to the hard-coded. If we want something generic, when we create the object we could register what is the 'fallback' context.

Reply:
I'm using RHEL 7.9 with `3.10.0-1160.2.2.el7.x86_64` kernel and I do not have this symlink as well. I only have `/proc/self`. I could not find out why.
But this may be platform dependent, so using a different way to get this info is more appropriate? I guess that fall-backing is also an option and explaining why. Maybe it is just not in RHEL/CentOS 7?

Reply:
Right, I've investigated and it seems the `thread-self` symlink only exists since kernel >= 3.17 . 

I'll manually construct the path - maybe that works also for kind nodes. If not, I'll fallback to a default.